src/logic/sourcemap.ts

```diff
@@ -0,0 +1,67 @@
+import * as vlq from 'vlq';
+
+export class SourceMap {
+  version: number;
+  sources: string[];
+  names: string[];
+  mappings: string;
+
+  pcToLine: { [key: number]: number };
+  lineToPc: { [key: number]: number[] };
+
+  constructor({
+    version,
+    sources,
+    names,
+    mappings,
+  }: {
+    version: number;
+    sources: string[];
+    names: string[];
+    mappings: string;
+  }) {
+    this.version = version;
+    this.sources = sources;
+    this.names = names;
+    this.mappings = mappings;
+
+    if (this.version !== 3)
+      throw new Error(`Only version 3 is supported, got ${this.version}`);
+
+    if (this.mappings === undefined)
+      throw new Error(
+        'mapping undefined, cannot build source map without `mapping`'
+      );
+
+    const pcList = this.mappings.split(';').map((m) => {
+      const decoded = vlq.decode(m);
+      if (decoded.length > 1) return decoded[2];
+      return undefined;
+    });
+
+    this.pcToLine = {};
+    this.lineToPc = {};
+
+    let lastLine = 0;
+    for (const [pc, lineDelta] of pcList.entries()) {
+      // If the delta is not undefined, the lastLine should be updated with
+      // lastLine + the delta
+      if (lineDelta !== undefined) {
+        lastLine += lineDelta;
+      }
+
+      if (!(lastLine in this.lineToPc)) this.lineToPc[lastLine] = [];
+
+      this.lineToPc[lastLine].push(pc);
+      this.pcToLine[pc] = lastLine;
+    }
+  }
+
+  getLineForPc(pc: number): number {
```

Comment:
@barnjamin Friendly visual reminder to implement an analog of https://github.com/algorand/py-algorand-sdk/pull/353#discussion_r910177370 here.  

From my _very_ limited Typescript knowledge, it seems the return type can be `string | undefined` to encode optionality.

Reply:
Sorry - I meant `number | undefined` in my prior note.